src/companion.rs

```diff
@@ -0,0 +1,136 @@
+use anyhow::Context;
+use regex::Regex;
+use tokio::process::Command;
+
+use crate::github_bot::GithubBot;
+
+pub async fn companion_update(
+	github_bot: &GithubBot,
+	owner: &str,
+	repo: &str,
+	branch: &str,
+) -> anyhow::Result<()> {
+	let token = github_bot.client.auth_key().await?;
+	Command::new("rustup")
+		.arg("update")
+		.spawn()
+		.context("spawn rustup update")?
+		.await
+		.context("rustup update")?;
+	Command::new("git")
+		.arg("clone")
+		.arg("-vb")
+		.arg(branch)
+		.arg(format!(
+			"https://x-access-token:{token}@github.com/{owner}/{repo}.git",
+			token = token,
+			owner = owner,
+			repo = repo,
+		))
+		.arg("repo")
+		.spawn()
+		.context("spawn git clone")?
+		.await
+		.context("git clone")?;
+	Command::new("cargo")
+		.arg("update")
+		.arg("-vp")
+		.arg("sp-io")
+		.current_dir("./repo")
+		.spawn()
+		.context("spawn cargo update")?
+		.await
+		.context("cargo update")?;
+	Command::new("git")
+		.arg("commit")
+		.arg("-a")
+		.arg("-m")
+		.arg("'Update substrate'")
+		.current_dir("./repo")
+		.spawn()
+		.context("spawn git commit")?
+		.await
+		.context("git commit")?;
+	Command::new("git")
+		.arg("push")
+		.arg("-v")
+		.current_dir("./repo")
+		.spawn()
+		.context("spawn git push")?
+		.await
+		.context("git push")?;
+	Command::new("rm")
+		.arg("-rf")
+		.arg("repo")
+		.spawn()
+		.context("spawn rm")?
+		.await
+		.context("rm")?;
+	Ok(())
+}
+
+pub fn companion_parse(body: &str) -> Option<(String, String, String, i64)> {
+	let re = Regex::new(
+		r"companion.*(?P<html_url>https://github.com/(?P<owner>[[:alpha:]]+)/(?P<repo>[[:alpha:]]+)/pull/(?P<number>[[:digit:]]+))"
```

Comment:
unfortunately for some reasons it could also be that the companion reference will lack the https:/github.com part. see: https://github.com/paritytech/substrate/blob/master/.maintain/gitlab/check_polkadot_companion_status.sh#L25
on top of that the [check_polkadot_companion_build](https://github.com/paritytech/substrate/blob/master/.maintain/gitlab/check_polkadot_companion_build.sh#L33) ci check also includes the feature to look for a branch on polkadot with the exact same name and treat it as its companion.
for conistency reasons would that be possible to add?

Reply:
ah thanks, I wasn't aware of those.  I will try to match all of them.

Reply:
This should match both additional cases you mentioned.  the shorter form will be matched by `companion_parse_short`.  If we can't find a direct link in the pr description we check for a matching branch name [here](https://github.com/paritytech/parity-processbot/pull/201/files#diff-074e70b14c7ade421c7b5c20d636ff07R1324) (using [this](https://github.com/paritytech/parity-processbot/pull/201/files#diff-527c6879361b080ebbabaea7e11ab989R35)).